pkg/gateway/types/users.go

```diff
@@ -36,12 +36,18 @@ type User struct {
 	OriginalUsername string     `json:"-"`
 }
 
-func ConvertUser(u *User, roleFixed bool, authProviderName string) *types2.User {
+func ConvertUser(u *User, roleFixed bool, authProviderName string, effectiveRole ...types2.Role) *types2.User {
 	if u == nil {
 		return nil
 	}
 
-	groups := u.Role.Groups()
+	// Determine which role to use for groups
+	role := u.Role
+	if len(effectiveRole) > 0 {
+		role = effectiveRole[0]
+	}
```

Comment:
I understand what is happening here, but this is the type of thing that gets lost or forgotten about over time.

My preference would be to have two functions: `ConvertUser` and `ConvertUserWithEffectiveRole` (this would have signature `ConvertUserWithEffectiveRole(u *User, roleFixed bool, authProviderName string, effectiveRole types2.Role) *types2.User`.

The `ConvertUser` would just call `ConverUserWithEffectiveRole` and pass `u.Role` for the `effectiveRole`.

Reply:
Good call